model/map/src/main/java/org/keycloak/models/map/singleUseObject/MapSingleUseObjectProvider.java

```diff
@@ -65,121 +65,69 @@ private ActionTokenValueModel singleUseEntityToAdapter(MapSingleUseObjectEntity
     }
 
     @Override
-    public void put(ActionTokenKeyModel actionTokenKey, Map<String, String> notes) {
-        if (actionTokenKey == null || actionTokenKey.getUserId() == null || actionTokenKey.getActionId() == null || actionTokenKey.getActionVerificationNonce() == null) {
-            return;
-        }
-
-        LOG.tracef("put(%s, %s, %s)%s", actionTokenKey.getUserId(), actionTokenKey.getActionId(), actionTokenKey.getActionVerificationNonce(), getShortStackTrace());
-
-        DefaultModelCriteria<ActionTokenValueModel> mcb = criteria();
-        mcb = mcb.compare(ActionTokenValueModel.SearchableFields.USER_ID, ModelCriteriaBuilder.Operator.EQ, actionTokenKey.getUserId())
-                .compare(ActionTokenValueModel.SearchableFields.ACTION_ID, ModelCriteriaBuilder.Operator.EQ, actionTokenKey.getActionId())
-                .compare(ActionTokenValueModel.SearchableFields.ACTION_VERIFICATION_NONCE, ModelCriteriaBuilder.Operator.EQ, actionTokenKey.getActionVerificationNonce().toString());
+    public void put(String key, long lifespanSeconds, Map<String, String> notes) {
+        // try to parse provided key to check if we are storing an action token
+        ActionTokenKeyModel actionTokenKey = DefaultActionTokenKey.from(key);
```

Comment:
`MapSingleUseObjectProvider` should not need to distinguish between action-token and other objects. Ideally, we should completely remove `DefaultActionTokenKey` and `ActionTokenValueModel` from this class. Would this be possible? 

This way, we could remove all ActionToken-related fields from the storage and retain only `id`, `objectKey` and `notes`. Do you see any problem with this?

Reply:
I removed the distinction between action tokens and single-use objects from  `MapSingleUseObjectProvider`. Currently I'm using id for action tokens from upper layers in format which is defined by serialization of an action token. The format is as follow: `encodedUserId.expiration.actionVerificationNonce.actionId` We need to agree on the format.
This allowed me to remove `userId`, `actionId`, `actionVerificationNonce` fields from everywhere.

Now question is whether we can go even further and remove also `objectKey` and all custom `SingleUseObjectModelCriteriaBuilder|SingleUseObjectKeycloakTransaction|SingleUseObject*Storage`?

Reply:
Thank you @martin-kanis, it looks much cleaner now, and I would say it makes sense to have it this way. 

> Now question is whether we can go even further and remove also objectKey and all custom SingleUseObjectModelCriteriaBuilder|SingleUseObjectKeycloakTransaction|SingleUseObject*Storage?

We can't get rid of `objectKey` completely because JPA is using it. On the other hand, we could move the field to JPA only. Currently, it seems we are having a field, necessary only in JPA available to all other implementations. This probably doesn't need to be solved here, I would prefer to discuss and possibly resolve within another issue. See my other comments for more details.